File: functions/javascript/olivia.mjs
import * as readline from "node:readline/promises";
import { GoogleGenerativeAI } from "@google/generative-ai";
import fs from "fs";
import Database from 'better-sqlite3';

const saveRecipeDeclaration = {
    name: "saveRecipe",
    parameters: {
        type: "OBJECT",
        description: "Save a new recipe",
        properties: {
            name: {
                type: "STRING",
                description: "Name of the recipe.",
            },
            ingredients: {
                type: "STRING",
                description: "List of ingredients needed to prepare the recipe, with their quantities.",
            },
            instructions: {
                type: "STRING",
                description: "List of specific steps required to prepare the recipe, including order, cooking times, and any details that need to be specified so that a person with no prior knowledge can successfully make it.",
            },
        },
        required: ["name", "ingredients", "instructions"],
    }
};

const listRecipesDeclaration = {
    name: "listRecipes"
}

let tableExists = false

function listRecipes() {
    let recipes, db;
    try {
        db = new Database('recipes.db');
        const stmt = db.prepare("select name, ingredients, instructions from recipe");
        recipes = stmt.all();
        tableExists = true;
    } catch {
        recipes = "There are no recipes.";
    } finally {
        db.close();
    }
    return recipes;
}

function saveRecipe(name, ingredients, instructions) {
    let result, db;
    try {
        db = new Database('recipes.db');
        if (!tableExists) {
            const stmtTabla = db.prepare(`CREATE TABLE if not exists recipe (
                [name] TEXT NOT NULL,
                [ingredients] TEXT,
                [instructions] TEXT)`);
            stmtTabla.run();
        }
        const stmtInsert = db.prepare("insert into recipe (name, ingredients, instructions) values (?,?,?)");
        stmtInsert.run(name, ingredients, instructions);
        result = "OK";
    } catch {
        result = "Error saving recipe";
    } finally {
        db.close();
    }
    return result;
}

class GeminiClient {
    constructor(model, systemPrompt) {
        this.model = model;
        const genai = new GoogleGenerativeAI(process.env.GOOGLE_API_KEY);
        let genModel = genai.getGenerativeModel(
            {
                model: model,
                safetySettings: [
                    {
                        "category": "HARM_CATEGORY_HARASSMENT",
                        "threshold": "BLOCK_NONE"
                    },
                    {
                        "category": "HARM_CATEGORY_HATE_SPEECH",
                        "threshold": "BLOCK_NONE"
                    },
                    {
                        "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
                        "threshold": "BLOCK_NONE"
                    },
                    {
                        "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
                        "threshold": "BLOCK_NONE"
                    },
                ],
                tools: {
                    functionDeclarations: [saveRecipeDeclaration, listRecipesDeclaration],
                },
            }, { apiVersion: 'v1beta', });
        this.chatSession = genModel.startChat();
        this.chatSession.sendMessage(systemPrompt); // we don't display the response, so don't need to handle the promise
    }

    async _chat(functionName, functionResponse) {
        const ret = await this.chatSession.sendMessage([{ "functionResponse": { "name": functionName, "response": { recipes: functionResponse } } }]);
        return ret.response.text();
    }

    async chat(userMessage) {
        const resp = await this.chatSession.sendMessage(userMessage);
        let modelMessage;
        if (resp.response.functionCall()) {
            const fc = resp.response.functionCall();
            console.log("-- function call: " + fc.name); // To be able to tell if function is actually being called
            if (fc.name === "listRecipes") {
                modelMessage = await this._chat(fc.name, listRecipes());
            } else if (fc.name === "saveRecipe") {
                modelMessage = await this._chat(fc.name, saveRecipe(fc.args.name, fc.args.ingredients, fc.args.instructions));
            } else {
                modelMessage = "Function call!! " + fc.name; // Should never happen
            }
        } else modelMessage = resp.response.text();
        return modelMessage;
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let client = new GeminiClient("gemini-1.0-pro", fs.readFileSync("olivia.md", "utf8"));
let userText;
while (userText = (await rl.question('What do you want to say to Olivia?\n> ')).trim()) {
    console.log(await client.chat(userText));
}
rl.close();

File: function_calling_anthropic/python/olivia.py
from anthropic import Anthropic
import sqlite3
import io
import json

save_recipe_declaration = {
    "name": "save_recipe",
    "description": "Save a new recipe",
    "input_schema": {
        "type": "object",
        "properties": {
            "name": {
                "type": "string",
                "description": "Name of the recipe.",
            },
            "ingredients": {
                "type": "string",
                "description": "List of ingredients needed to prepare the recipe, with their quantities.",
            },
            "instructions": {
                "type": "string",
                "description": "List of specific steps required to prepare the recipe, including order, cooking times, and any details that need to be specified so that a person with no prior knowledge can successfully make it.",
            },
        },
        "required": ["name", "ingredients", "instructions"],
    }
}

list_recipes_declaration = {
    "name": "list_recipes",
    "description": "List saved recipes.",
    "input_schema": {
        "type": "object",
        "properties": {}
    }
}

table_exists = False

def list_recipes():
    # Gets the list of stored recipes
    conn = sqlite3.connect("recipes.db")
    try:
        recipes = [{"name":row[0], "ingredients":row[1], "instructions":row[2]} for row in conn.execute("select name, ingredients, instructions from recipe")]
    except:
        recipes = []
    conn.close()
    return recipes

def save_recipe(name, ingredients, instructions):
    # Save a new recipe
    conn = sqlite3.connect("recipes.db")
    cur = conn.cursor()
    if not table_exists:
        cur.execute("""CREATE TABLE if not exists recipe (
                    [name] TEXT NOT NULL,
                    [ingredients] TEXT,
                    [instructions] TEXT)""")
    cur.execute("insert into recipe (name, ingredients, instructions) values (?,?,?)",
                (name, ingredients, instructions))
    conn.commit()
    cur.close()
    conn.close()

class ClaudeClient:
    def __init__(self, model, system_prompt):
        self.model = model
        self.system_prompt = system_prompt
        self.anthropic = Anthropic()
        self.messages = []

    def __call_function(self, llf):
        if llf.name == "list_recipes":
            return json.dumps(list_recipes())
        elif llf.name == "save_recipe":
            return json.dumps(save_recipe(llf.input["name"], llf.input["ingredients"], llf.input["instructions"]))


    def chat(self, user_message):
        self.messages.append(
            {
                "role": "user",
                "content": user_message
            }
        )
        response = self.anthropic.messages.create(
            max_tokens=4096,
            system=self.system_prompt,
            messages=self.messages,
            model = self.model,
            tools=[save_recipe_declaration, list_recipes_declaration]
        )
        self.messages.append(
             {
                "role": "assistant",
                "content": response.content
            }
        )
        if response.stop_reason == "tool_use":
            output = ""
            function_result = []
            for c in response.content:
                if c.type == "text":
                    output = c.text
                elif c.type == "tool_use":
                    function_result.append(
                        {
                            "type": "tool_result",
                            "tool_use_id": c.id,
                            "content": self.__call_function(c)
                        })
            self.messages.append(
                {
                    "role": "user",
                    "content": function_result
                })
            response2 = response = self.anthropic.messages.create(
                max_tokens=4096,
                system=self.system_prompt,
                messages=self.messages,
                model = self.model,
                tools=[save_recipe_declaration, list_recipes_declaration]
            )
            self.messages.append(
                {
                    "role": "assistant",
                    "content": response2.content
                }
            )
            return output+"\n"+response2.content[0].text
        
        return response.content[0].text

with io.open("olivia.md", "r") as arch:
    system_prompt = "\n".join(arch.readlines())

client = ClaudeClient("claude-3-5-sonnet-20240620", system_prompt)
while (user_text := input("What do you want to say to Olivia?\n> ").strip()) !="":
    if len(user_text)==0: 
        break
    print(client.chat(user_text))

